Cpp_hands_on_programming/lib/pipe/Ipc_pipe.cc

```diff
@@ -35,55 +50,103 @@ void Ipc_pipe::send() {
 }
 
 void Ipc_pipe::receive() {
     std::cout << "Starting pipeReceive.." << std::endl;
 	int fdp, bytes;
-	off_t fileSize(0), total(0);
+	off_t fileSize(0), total(0), sendFileSize(0);
+	bool headerSent(false);
 	std::vector<char> buffer(PIPE_BUF);
 	// open pipe 
-	while ((fdp = open(name.c_str(), O_RDONLY)) == -1) {
-		std::this_thread::sleep_for (std::chrono::seconds(1));
-	}
+	t.startTimer();
+	do {
+		errno = 0;
+		fdp = open(name.c_str(), O_RDONLY | O_NONBLOCK); 
+		t.checkTimer("Cannot connect to pipeSend. Errno: " + std::string(strerror(errno)), "Connecting pipeSend..");
+	} while(errno == ENOENT);
+
+	std::cout << "Connected!" << std::endl;
 	// Start read and write loop
-	while ((bytes = read(fdp, buffer.data(), PIPE_BUF)) != 0) {	
-		if (bytes == -1) {
-			throw(std::runtime_error("Ipc_pipe::receive: read"));
-		}
+	do {
+		t.startTimer();
+		do {
+			errno = 0;
+			bytes = read(fdp, buffer.data(), PIPE_BUF);
+			if (bytes == -1) {
+				if (errno == EAGAIN) {
+					t.checkTimer("Pipe is empty for too long. Errno: " + std::string(strerror(errno)));
+				} else {
+					throw(std::runtime_error("Ipc_pipe::receive: read. Errno: " + std::string(strerror(errno))));
+				}
+			}
+			else if (!headerSent) {
+				if (bytes == 0) {
+					t.checkTimer("Takes too long for write end to open");
+				} else {
+					auto sendFileName = std::string(buffer.begin(), buffer.begin() + bytes);
+					if ((sendFileSize = getFileSize(sendFileName)) == -1) {
+						throw(std::runtime_error("Ipc_pipe::receive: getFilesize-sendFile. Errno: " + std::string(strerror(errno))));
+					}
+					headerSent = true;
+					total -= bytes;
+				}
+			}
+		} while (errno == EAGAIN || !headerSent);
 		// open and write to file
-		buffer.resize(bytes);
-		if (writeFile(filename, total, buffer) == -1) {
-			throw(std::runtime_error("Ipc_pipe::receive: writeFile"));
+		
+		if (total >= 0) {
+			buffer.resize(bytes);
+			if (writeFile(filename, total, buffer) == -1) {
+				throw(std::runtime_error("Ipc_pipe::receive: writeFile"));
+			}
 		}
 		total += bytes;
-	}
-	
+		
+	} while (bytes != 0);
+
 	if ((fileSize = getFileSize(filename)) == -1) {
-		throw(std::runtime_error("Ipc_method:: getFileSize"));
+		throw(std::runtime_error("Ipc_pipe::receive: getFileSize"));
 	}
-	if (total == fileSize) {
+	if (fileSize == sendFileSize) {
 		std::cout << "Write to file from pipe successfully, exiting the program.." << std::endl;
 	} else {
-		throw(std::runtime_error("Ipc_pipe::receive: fileSize and total read difference"));
+		throw(std::runtime_error("Ipc_pipe::receive: Received file and send file different"));
 	}
 
 	close(fdp);
 }
 
 Ipc_pipe::~Ipc_pipe() {
 	unlink(name.c_str());
+}
+
+void Ipc_pipe::sendHeader(int fd) {
+	if (write(fd, filename.c_str(), filename.length()) == -1) {
+		throw(std::runtime_error("Ipc_pipe::sendHeader fail!"));
+	}
+	// wait for receiver to retrieve the header
+	std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
Use sleep_for for synchronisation is not the best thing to do in my opinion.

Reply:
I was thinking of using a second pipe used for transferring header first, but it would add more overhead, I think this workaround way works fine for most of the cases and less expensive compared to creating one more pipe. Do you have any suggestions to do it in a cheaper way?

Reply:
You can send into the pipe regular size data (regular buffer size) and put into that data a null terminate char '\0' to say it's the end of the filename

Reply:
Hmm, it could work but it would involve many C code, I would prefer doing it in a more C++ way

Reply:
I don't think it involve "many" C code.
```c++
std::string string = "awfawfa";
string.c_str(); //will terminate by a \0

```

But as you wish